### PR TITLE
[Docs] Deployment on existing infra

### PR DESCRIPTION
**Pull request by romilbhardwaj**
This PR adds a deployment script to help users easily deploy k8s + SkyPilot given a list of SSH IPs.

### CODE REVIEW

**Comment by romilbhardwaj** [comment]
Updated with figure and general polishing.

File: docs/source/images/sky-existing-infra-workflow.png
Comment (Michaelvll) [inline docs/source/images/sky-existing-infra-workflow.png:1]: nit: `deploy.sh` may feel a bit janky to some people. Will it be hard to turn it into a CLI or part of `sky local up`?

**Comment by romilbhardwaj** [comment]
Updated to integrate the script into `sky local up --ip` CLI.

**Comment by romilbhardwaj** [comment]
Updated with a polished guide and minor fixes.

File: docs/source/reservations/existing-clusters.rst [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+      $ sky launch --cloud kubernetes --gpus H100:1 -- nvidia-smi
+
+   .. tip::
+
+     You can also use ``kubectl`` to interact and perform administrative operations on the cluster.
+
+Cleanup
+-------
+
+To clean up all state created by SkyPilot on your cluster, use the ``--cleanup`` flag:
```
Comment (concretevitamin) [inline docs/source/reservations/existing-clusters.rst:134]: Reading "all state" made me wonder what are those states? How about adding a quick section with a bullet list / snippet showing this installed state? Users have asked "what do you install on my cluster".

**Review by concretevitamin — COMMENTED** [review]
Nice @romilbhardwaj! Took a pass over the docs.

File: docs/source/reservations/existing-clusters.rst [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+   You can now use SkyPilot to launch your :ref:`development clusters <dev-cluster>` and :ref:`training jobs <ai-training>` on your own infrastructure.
+
+   .. code-block:: console
+
+      $ sky show-gpus --cloud kubernetes
+      Kubernetes GPUs
+      GPU   QTY_PER_NODE  TOTAL_GPUS  TOTAL_FREE_GPUS
+      L4    1, 2, 4       12          12
+      H100  1, 2, 4, 8    16          16
+      Kubernetes per node GPU availability
```
[the remark this answers is not in the document]
Comment (romilbhardwaj) [reply docs/source/reservations/existing-clusters.rst:116]: Ah our code does have a new line, somehow that didn't get copied here. fixed.

File: docs/source/reservations/existing-clusters.rst [hunk 3]
```diff
@@ -0,0 +1,143 @@
+.. _existing-clusters:
+
+Deploy SkyPilot on existing clusters
```
[the remark this answers is not in the document]
Comment (romilbhardwaj) [reply docs/source/reservations/existing-clusters.rst:3]: Using "local" or "on-prem" may make a user think this is not applicable to machines in the cloud. We have had several users have this existing cluster use case when their cluster was from a cloud provider. 

Also, using "local" or "on-prem" does not solve the "existing k8s" concern - both can be assumed to have k8s.

We can consider changing to "existing clusters" to "existing machines" if that helps.

File: docs/source/images/sky-existing-infra-workflow-light.png
[the remark this answers is not in the document]
Comment (romilbhardwaj) [reply docs/source/images/sky-existing-infra-workflow-light.png:1]: * Good point, node conflicts with physical nodes. Changed to pods.
* I think putting a single arrow de-emphasizes the need for having something like SkyPilot and makes it seem like the users are having to deal with low-level infra themselves.

File: docs/source/reservations/existing-machines.rst [hunk 4]
```diff
@@ -70,22 +68,22 @@ Deploying SkyPilot
...
+2. Run ``sky local up`` and pass the ``ips.txt`` file, SSH username, and SSH key as arguments:
 
    .. code-block:: bash
 
       IP_FILE=ips.txt
-      SSH_USERNAME=username
+      SSH_USER=username
       SSH_KEY=path/to/ssh/key
-      sky local up --ip $IP_FILE --username $SSH_USERNAME --key-path $SSH_KEY
+      sky local up --ip $IP_FILE --ssh-user SSH_USER --ssh-key-path $SSH_KEY
```
Comment (concretevitamin) [inline docs/source/reservations/existing-machines.rst:78]: Is it `--ip` or `--ips` (L86)? Latter is more intuitive. (Or --ipfile like DeepSpeed's --hostfile.)

File: sky/cli.py [hunk 5]
```diff
@@ -5206,6 +5198,123 @@ def local_up(gpus: bool):
             f'{gpu_hint}')
 
 
+def deploy_remote_cluster(ip_file, ssh_user, ssh_key_path, cleanup):
```
Comment (concretevitamin) [inline sky/cli.py:5201]: - make private
- type hints?

File: sky/utils/log_utils.py [hunk 6]
```diff
@@ -129,6 +129,75 @@ def __exit__(self, except_type, except_value, traceback):
         self.status_display.stop()
 
 
+class SkyRemoteUpLineProcessor(LineProcessor):
```
Comment (concretevitamin) [inline sky/utils/log_utils.py:132]: type hint the member funcs?

File: docs/source/reservations/existing-machines.rst (hunk 4 above)
Comment (romilbhardwaj) [reply docs/source/reservations/existing-machines.rst:78]: ah good catch - it is indeed `--ips`, this typo was a remnant from the old deploy.sh.

File: sky/utils/log_utils.py [hunk 7]
```diff
@@ -1,6 +1,7 @@
 """Logging utils."""
 import enum
-from typing import List, Optional
+from types import TracebackType
```
Comment (concretevitamin) [inline sky/utils/log_utils.py:3]: styleguide: import module only https://google.github.io/styleguide/pyguide.html#2241-exemptions

File: docs/source/reservations/existing-machines.rst [hunk 8]
```diff
@@ -3,7 +3,7 @@
 Deploy SkyPilot on existing machines
 ====================================
 
-This guide will help you deploy SkyPilot on your existing machines - whether they are on-premisesc or reserved instances on a cloud provider.
+This guide will help you deploy SkyPilot on your existing machines - whether they are on-premises or reserved instances on a cloud provider.
```
Comment (concretevitamin) [inline docs/source/reservations/existing-machines.rst:6]:
```suggestion
This guide will help you deploy SkyPilot on your existing machines — whether they are on-premises or reserved instances on a cloud provider.
```

File: docs/source/reservations/existing-machines.rst [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+* SSH access from local machine to all remote machines with key-based authentication and passwordless sudo
+* All machines must use the same SSH key and username
+* All machines must have network access to each other
+* Port 6443 must be accessible on at least one node from your local machine
+
+Deploying SkyPilot
+------------------
+
+1. Create a file ``ips.txt`` with the IP addresses of your machines with one IP per line.
+   The first node will be used as the head node - this node must have port 6443 accessible from your local machine.
```
Comment (concretevitamin) [inline docs/source/reservations/existing-machines.rst:59]:
```suggestion
   The first node will be used as the head node — this node must have port 6443 accessible from your local machine.
```

**Review by concretevitamin — APPROVED** [review]

**Pull request merged**